### ISSUE TITLE
Temporary test files are not always cleaned up

### ISSUE THREAD

**Issue by phile314** [agda/agda#2317]
Sometimes, the compiler tests leave temporary files behind which should have been cleaned up. This seems to happen when aborting the test run with CTRL-C.

The problem could be that the test runner launches Agda sub processes which are not killed when the test driver itself is killed. Agda then may (re-)create the temporary output directories, which are no longer cleaned up as the test runner itself is already dead.

**Comment by andreasabel** [comment]
Any update on this?  How about creating the temporary files in the TEMP folder of the system instead of locally?  Then we do not have to worry about cleaning them up always.

Another take would be that the test runner cleans up everything that is left over from previous runs as well.

**Comment by nad** [comment]
I've noticed this bug when running the LaTeX and HTML tests. The [process documentation](https://hackage.haskell.org/package/process-1.5.0.0/docs/System-Process.html#g:4) seems to contain relevant information.

**Comment by phile314** [comment]
I investigated this a few months ago but didn't find the root cause. The tests use ``withTempDirectory``, which should always remove the test directory [1]. For some reason, this does not seem to work. 

Maybe @andreasabel suggestion is the better way to go. One option would be to have one temp directory, e.g. "test-temp" in the project root. Cleanup could be done on the start of the test run. This would also have the advantage that one could look at the generated temp files from the last run when a test failed, whereas now the temp files are deleted immediately when a test fails.


[1] https://github.com/agda/agda/blob/master/test/LaTeXAndHTML/Tests.hs#L81

**Comment by nad** [comment]
> For some reason, this does not seem to work.

Did you try setting `delegate_ctlc` to `True`? (See the  [process documentation](https://hackage.haskell.org/package/process-1.5.0.0/docs/System-Process.html#g:4).)

**Comment by nad** [comment]
> Did you try setting delegate_ctlc to True?

I tried this (using `-halt-on-error` rather than `-interaction=errorstopmode`). Result: When I interrupt the LaTeX program I get a diff between the golden value and the LaTeX error message. The process documentation states that "in delegate_ctlc mode, waitForProcess and getProcessExitCode will throw a UserInterrupt exception if the process terminated with ExitFailure (-SIGINT)". If Agda is interrupted, then the exitcode is 130 (using my setup), which I guess is the same as -SIGINT. However, when I interrupt `lualatex`, `pdflatex` or `xelatex` the exit code is 1. Thus `delegate_ctlc` doesn't quite work for us. However, we could presumably detect Ctrl-C in the test runner, wait until all the external processes have terminated, run the cleanup code, and then exit (with -SIGINT as the exit code).

Given that the tests are invoked by Tasty, and that Tasty uses multiple threads, it would perhaps be better if this logic was implemented in Tasty rather than in our test runner.
